DISTMYSQL-242 Document replication manager script (#64)
percona/pdmysql-docs: docs/replication-manager/replication-manager-for-pxc.md

# Replication manager for Percona XtraDB Cluster

> The replication manager script helps manage source-source replication between multiple Percona XtraDB Cluster clusters.

Reviewer: What is source-source replication? perhaps it makes sense to explain it either here, or add a new entry to glossary and reference to it from this doc

Author: Changed to multi-source and added to Glossary.

> ## Preconditions

Reviewer 2: Should this be Example?
Preconditions imply that we should tell them what they need beforehand.

Author: Updated

> For example, we deploy source-source replication links between three data centers: DC1, DC2, and DC3. In each data center, three Percona XtraDB Cluster nodes form distinct Galera clusters. The data centers contain the following nodes:

Reviewer: ```suggestion
For example, we deploy source-source replication links between three data centers: DC1, DC2, and DC3. In each data center, there are three Percona XtraDB Cluster nodes form distinct Galera clusters. The data centers contain the following nodes:
```

Author: Updated

|  DC1  |  DC2  |  DC3  |
| ----- | ----- | ----- |
| DC1-1 | DC2-1 | DC3-1 |
| DC1-2 | DC2-2 | DC3-2 |
| DC1-3 | DC2-3 | DC3-3 |
 
The topology is the following:

```text
DC2 <=> DC1 <=> DC3
```

DC1 is a slave of DC2 and DC3. DC2 and DC3 are slaves of DC1.

The example of the minimal MySQL configuration file for DC1 that is common for all three DC1 nodes:

```text
[mysqld]
# General galera reqs
default_storage_engine=InnoDB
innodb_autoinc_lock_mode=2

# Replication settings
binlog_format=ROW
server-id=1
log-bin=mysql-bin
log_slave_updates
expire_logs_days=7
gtid_mode = ON
enforce_gtid_consistency=ON
master_info_repository = TABLE
relay_log_info_repository = TABLE
skip-slave-start
    
# Galera configuration
wsrep_provider=/usr/lib/galera3/libgalera_smm.so
wsrep_cluster_address=gcomm://10.0.4.160,10.0.4.162,10.0.4.163
wsrep_slave_threads= 2
wsrep_log_conflicts
wsrep_cluster_name=DC1
pxc_strict_mode=ENFORCING
wsrep_sst_method=xtrabackup-v2
wsrep_sst_auth="root:root"
```

> All nodes have the same `server-id` value. The repositories are set to `TABLE` because the multi-source replication syntax is used. This means a node can become a slave of more than one remote cluster.

Reviewer 2: Multisource replication allows a node to become a replica in more than one remote cluster.

Reviewer 3: yes fine

Author: Updated

> All nodes have the same `server-id` value. The repositories are set to `TABLE` because the multi-source replication syntax is used. Multi-source replication allows a node to become a replica in more than one remote cluster.

Reviewer 2: In a multi-source replication topology, the replicas require the `master_info_repository` and `relay_log_info_repository` values to be TABLE. These settings cannot be `FILE`.

Author: Updated

      `host` varchar(40) NOT NULL,
      `weight` int(11) NOT NULL DEFAULT 0,
      `localIndex` int(11) DEFAULT NULL,
      `isSlave` enum('No','Yes','Proposed','Failed') DEFAULT 'No',
      `lastUpdate` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
      `lastHeartbeat` timestamp NOT NULL DEFAULT '1971-01-01 00:00:00',
      `connectionName` varchar(64) NOT NULL,
      PRIMARY KEY (`connectionName`,`host`),
      KEY `idx_host` (`host`)
    ) ENGINE=InnoDB DEFAULT CHARSET=latin1;
    CREATE TABLE `cluster` (
      `cluster` varchar(31) NOT NULL,
      `masterCandidates` varchar(255) NOT NULL,
      `replCreds` varchar(255) NOT NULL,
      PRIMARY KEY (`cluster`)
    ) ENGINE=InnoDB DEFAULT CHARSET=latin1;
    CREATE TABLE `link` (
      `clusterSlave` varchar(31) NOT NULL,
      `clusterMaster` varchar(31) NOT NULL,
      PRIMARY KEY (`clusterSlave`,`clusterMaster`)
    ) ENGINE=InnoDB DEFAULT CHARSET=latin1;
    CREATE TABLE `weight` (
     `cluster` varchar(31) NOT NULL,
     `nodename` varchar(255) NOT NULL,
     `weight` int NOT NULL DEFAULT 0, 
     PRIMARY KEY (`cluster`,`nodename`)
    ) ENGINE=InnoDB DEFAULT CHARSET=latin1;
    ```

>     The replication table is written to by the replication manager. You do not need to adjust the table.

Reviewer 2: is written to by?
The replication manager writes to the replication table.

Author: Updated

    INSERT INTO `cluster` VALUES ('DC2','10.0.4.164 10.0.4.165 10.0.4.166','master_user=\'repl\', master_password=\'replpass\'');
    INSERT INTO `cluster` VALUES ('DC3','10.0.4.167 10.0.4.168 10.0.4.169','master_user=\'repl\', master_password=\'replpass\'');
    ```

    Add the required links:

    ```bash
    INSERT INTO `link` VALUES ('DC1','DC2');
    INSERT INTO `link` VALUES ('DC1','DC3');
    INSERT INTO `link` VALUES ('DC2','DC1');
    INSERT INTO `link` VALUES ('DC3','DC1');
    ```

    Add the weight:

    ```bash
    INSERT INTO `weight` VALUES('DC1','DC1-1',10); 
    INSERT INTO `weight` VALUES('DC1','DC1-2',11); 
    INSERT INTO `weight` VALUES('DC2','DC2-1',9);
    INSERT INTO `weight` VALUES('DC2','DC2-2',12);
    INSERT INTO `weight` VALUES('DC3','DC3-1',11);
    ```

    The node in the cluster with the highest value is the first to be selected.

9. Provision the remote clusters and start replication. 

    On one of the DC1 nodes, for example DC1-1, perform a `mysqldump` with:

    ```bash
    [root@DC1-1 ~]# mysqldump -u root -p --master-data=2 --single-transaction -R -A -E > dump.sql
    ```  

> 10. Compress the file if it's too large. You can use Percona Xtrabackup to compress the data. Copy the backup file to one node in each remote cluster, for example to DC2-1 and DC3-1. 

Reviewer: ```suggestion
10. Compress the file if it's too large. You can use Percona XtraBackup to compress the data. Copy the backup file to one node in each remote cluster, for example to DC2-1 and DC3-1. 
```

Author: Updated

    To restore the dump run:

    ```bash
    [root@DC2-1 ~]# mysql -u root -p < dump.sql
    ```

    and

    ```bash
    [root@DC3-1 ~]# mysql -u root -p < dump.sql
    ```

> 11. Configuring the replication. Set up the first replication links manually. 

Reviewer: ```suggestion
11. Configure the replication. Set up the first replication links manually. 
```

Reviewer: You have infinitives everywhere. Using gerund here is inconsistent

Author: Updated

    For example, on DC2-1 run:

    ```bash

>     mysql> change master to master_host='WAN IP of DC1-1', master_user='repl', master_password='replpass', MASTER_AUTO_POSITION = 1 for channel 'DC2-DC1';

Reviewer 2: https://dev.mysql.com/doc/refman/8.0/en/change-replication-source-to.html

Author: Updated

    mysql> start slave for channel 'DC2-DC1';
    ```
    
    on DC3-1 run:

    ```bash

>     mysql> change master to master_host='WAN IP of DC1-1', master_user='repl', master_password='replpass', MASTER_AUTO_POSITION = 1 for channel 'DC3-DC1';

Reviewer 2: https://dev.mysql.com/doc/refman/8.0/en/change-replication-source-to.html

Author: Updated

    mysql> start slave for channel 'DC3-DC1';
    ```

    For the other direction, use DC1-1 for both:

    ```bash
    mysql> change master to master_host='WAN IP of DC2-1', master_user='repl', master_password='replpass', MASTER_AUTO_POSITION = 1 for channel 'DC1-DC2';

>     mysql> start slave for channel 'DC1-DC2';

Reviewer 2: https://dev.mysql.com/doc/refman/8.0/en/start-replica.html

Author: Updated

    mysql> change master to master_host='WAN IP of DC3-1', master_user='repl', master_password='replpass', MASTER_AUTO_POSITION = 1 for channel 'DC1-DC3';
    mysql> start slave for channel 'DC1-DC3';
    ```

    Now, all the clusters are linked in a source to source way. 

12. Pull in the `replication_manager.sh` script. On each node, perform the following steps:

    ```bash

>     # cd /usr/local/bin

Reviewer: Why is the code commented?

Author: Changed to $

>     $ cd /usr/local/bin

Reviewer 4: The replication script will be in pdpxc packages which means that there will not be any need to get it with wget. It will be installed with yum or apt

Reviewer 4: The steps:
1) download percona-release
2) enable pdpxcXX repo
3) install:
deb: sudo apt install  percona-replication-manager
rpm: sudo yum install  percona-replication-manager

No need to make chmod

Author: Updated

    # wget https://github.com/y-trudeau/Mysql-tools/raw/master/PXC/replication_manager.sh
    # chmod u+x replication_manager.sh
    ```

>     When executed for the first time, the replication manager will detect the current replication links and insert rows in to the `percona.replication table`. In order to avoid issues, start with the replica nodes. The mysql credentials should be specified in the `/root/.my.cnf` file. On these nodes (DC1-1, DC2-1 and DC3-1), execute the script manually:

Reviewer: ```suggestion
    When executed for the first time, the replication manager detects the current replication links and inserts rows in to the `percona.replication table`. In order to avoid issues, start with the replica nodes. The mysql credentials should be specified in the `/root/.my.cnf` file. On these nodes (DC1-1, DC2-1 and DC3-1), execute the script manually:
```

Author: Updated

    
    ```bash  

>     # /usr/local/bin/replication_manager.sh

Reviewer: Same here: why is the command commented?

Author: Changed to $

    ```

    The replication state should be unchanged and the `percona.replication` table should contain the following rows:

    ```bash  

>     mysql> select * from percona.replication;

Reviewer 2: SELECT * FROM

Author: Updated

    ```

2. On DC1-2 node, stop MySQL, delete the content of the datadir to force the SST, and start MySQL. Run the following commands as root or using sudo:

    ```bash
    [root@DC1-2 ~]# /etc/init.d/mysql stop
    [root@DC1-2 ~]# rm -rf /var/lib/mysql/*
    [root@DC1-2 ~]# /etc/init.d/mysql start
    ```

3. Once the SST of DC1-2 node is completed, proceed on DC1-3. Run the following commands as root or using sudo:

    ```bash
    [root@DC1-3 ~]# /etc/init.d/mysql stop
    [root@DC1-3 ~]# rm -rf /var/lib/mysql/*
    [root@DC1-3 ~]# /etc/init.d/mysql start
    ```

4. The DC1 cluster uses a single [GTID](../glossary.md#gtid) sequence. To make sure `GTID_PURGED` is the same, run the following commands on DC1-1, DC1-2, and DC1-3 nodes:

    ```bash
    mysql> flush logs;
    mysql> purge source logs to 'mysql-bin.000003';
    ```

    `mysql-bin.000003` is the last file returned from the show source logs. 

5. Adjust the aformentioned steps to set up the DC2 and DC3 clusters accordingly.
    
6. Add privileges for replication between all nodes only for DC1: 

    ```bash
    GRANT REPLICATION CLIENT, REPLICATION REPLICA ON *.* TO 'repl'@'%' identified by 'replpass';
    ```

> 7. Add the replication tables, the required links, and assign the weights. For example, we add the tables to DC1: 

Reviewer 2: Since you made them separate steps, you can remove the "required links and assign the weights". Those are not part of this step.

Author: Updated

    ```bash
    mysql> change replication source to source_host='WAN IP of DC2-1', source_user='repl', source_password='replpass', SOURCE_AUTO_POSITION = 1 FOR CHANNEL 'DC1-DC2';
    mysql> start replica FOR CHANNEL 'DC1-DC2';
    mysql> change replication source to source_host='WAN IP of DC3-1', source_user='repl', source_password='replpass', SOURCE_AUTO_POSITION = 1 FOR CHANNEL 'DC1-DC3';
    mysql> start replica FOR CHANNEL 'DC1-DC3';
    ```

    Now, all the clusters are linked in a source to source way. 

14. Pull in the `replication_manager.sh` script. On each node, perform the following steps:

    ```bash
    $ cd /usr/local/bin
    $ wget https://github.com/y-trudeau/Mysql-tools/raw/master/PXC/replication_manager.sh
    $ chmod u+x replication_manager.sh
    ```

    When executed for the first time, the replication manager detects the current replication links and inserts rows in to the `percona.replication table`. In order to avoid issues, start with the replica nodes. The mysql credentials should be specified in the `/root/.my.cnf` file. On these nodes (DC1-1, DC2-1 and DC3-1), execute the script manually:
    
    ```bash  

>     $ ./usr/local/bin/replication_manager.sh

Reviewer 4: The script location is `/usr/bin ` , so `/usr/bin/replication_manager.sh` or simply run command `replication_manager.sh`

Author: updated

    ??? example "Expected output"

        ```text
        +-------+--------+------------+-----------+---------------------+---------------------+----------------+
        | host  | weight | localIndex | isReplica | lastUpdate          | lastHeartbeat       | connectionName |
        +-------+------- +------------+-----------+---------------------+---------------------+----------------+
        | DC1-1 |      10|          1 | Yes       | 2017-06-30 13:03:01 | 2017-06-30 13:03:01 | DC1-DC2        |
        | DC1-1 |      11|          1 | Yes       | 2017-06-30 13:03:01 | 2017-06-30 13:03:01 | DC1-DC3        |
        | DC2-1 |       9|          1 | Yes       | 2017-06-30 13:03:01 | 2017-06-30 13:03:01 | DC2-DC1        |
        | DC3-1 |      11|          1 | Yes       | 2017-06-30 13:03:01 | 2017-06-30 13:03:01 | DC3-DC1        |
        +-------+--------+------------+-----------+---------------------+---------------------+----------------+
        12 rows in set (0.00 sec)
        ```

    If you have issues on these step, see the [Replication manager troubleshooting](replication-manager-troubleshooting.md) document.

    On these same nodes, enable the cron job:

    ```bash

>     * * * * * /usr/local/bin/replication_manager.sh 

Reviewer 4: the same here:
`/usr/bin/replication_manager.sh`

Author: updated

    mysql> change replication source to source_host='WAN IP of DC3-1', source_user='repl', source_password='replpass', SOURCE_AUTO_POSITION = 1 FOR CHANNEL 'DC1-DC3';
    mysql> start replica FOR CHANNEL 'DC1-DC3';
    ```

    Now, all the clusters are linked in a source to source way. 

14. Install and enable Percona XtraDB Cluster-based variant of Percona Distribution for MySQL as described in [Install Percona Distribution for MySQL](https://docs.percona.com/percona-distribution-for-mysql/8.0/installing.html#procedure).

15. Install the packages for `replication_manager.sh` script depending on your operating system. On each node, perform the following steps as root or using sudo:

    ```bash
    $ sudo apt install percona-replication-manager
    $ sudo yum install percona-replication-manager    
    ```

    When executed for the first time, the replication manager detects the current replication links and inserts rows in to the `percona.replication table`. In order to avoid issues, start with the replica nodes. The mysql credentials should be specified in the `/root/.my.cnf` file. On these nodes (DC1-1, DC2-1 and DC3-1), execute the script manually:
    
    ```bash  

>     $ ./usr/bin/replication_manager.sh

Reviewer 4: extra dot in the beginning

Author: removed

|  DC1  |  DC2  |  DC3  |
| ----- | ----- | ----- |
| DC1-1 | DC2-1 | DC3-1 |
| DC1-2 | DC2-2 | DC3-2 |
| DC1-3 | DC2-3 | DC3-3 |
 
The topology is the following:

```text
DC2 <=> DC1 <=> DC3
```

All of the nodes are sources and replicas. DC1 is a replica of DC2 and DC3. DC2 and DC3 are replicas of DC1.

```text
+----------------+---------------+
| clusterReplica | clusterSource |
+----------------+---------------+
| DC1            | DC2           |
| DC1            | DC3           |
| DC2            | DC1           |
| DC3            | DC1           |
+----------------+---------------+
```

The example of the minimal MySQL configuration file for DC1 that is common for all three DC1 nodes:

```text
[mysqld]
# General galera reqs
default_storage_engine=InnoDB
innodb_autoinc_lock_mode=2

# Replication settings
binlog_format=ROW
server-id=1
log-bin=mysql-bin
log_replica_updates
expire_logs_days=7
gtid_mode = ON
enforce_gtid_consistency=ON

> source_info_repository = TABLE

Reviewer 4: There is no source_info_repository variable. There is master_info_repository variable. I guess it should have been used here. But note that it is deprecated from 8.0.23.: 
https://dev.mysql.com/doc/relnotes/mysql/8.0/en/news-8-0-23.html . tables are the default for the replication metadata repositories in MySQL 8.0.
So I believe we can remove these ones from config. @Tusamarco ?

Reviewer 3: yes u can

Author: I have removed  master_info_repository and relay_log_info_repository variables since they are deprecated, and there are no options to replace them.

relay_log_info_repository = TABLE
skip-replica-start
    
# Galera configuration
wsrep_provider=/usr/lib/galera3/libgalera_smm.so
wsrep_cluster_address=gcomm://10.0.4.160,10.0.4.162,10.0.4.163
wsrep_replica_threads= 2
wsrep_log_conflicts
wsrep_cluster_name=DC1
pxc_strict_mode=ENFORCING
wsrep_sst_method=xtrabackup-v2
wsrep_sst_auth="root:root"
```

> All nodes have the same `server-id` value. In a multi-source replication topology, the replicas require the `source_info_repository` and `relay_log_info_repository` values to be `TABLE`. These settings cannot be `FILE`.

Reviewer 4: The same as in comment above

Reviewer 3: yes

Author: I have removed  master_info_repository and relay_log_info_repository variables since they are deprecated, and there are no options to replace them.

### Configuration steps

1. Bootstrap the cluster on DC1-1 node by running the following commands as root or using sudo:

    ```bash
    [root@DC1-1 ~]# /etc/init.d/mysql stop 
    [root@DC1-1 ~]# /etc/init.d/mysql bootstrap-pxc
    ```

2. On DC1-2 node, stop MySQL, delete the content of the datadir to force the SST, and start MySQL. Run the following commands as root or using sudo:

    ```bash
    [root@DC1-2 ~]# /etc/init.d/mysql stop
    [root@DC1-2 ~]# rm -rf /var/lib/mysql/*
    [root@DC1-2 ~]# /etc/init.d/mysql start
    ```

3. Once the SST of DC1-2 node is completed, proceed on DC1-3. Run the following commands as root or using sudo:

    ```bash
    [root@DC1-3 ~]# /etc/init.d/mysql stop
    [root@DC1-3 ~]# rm -rf /var/lib/mysql/*
    [root@DC1-3 ~]# /etc/init.d/mysql start
    ```

4. The DC1 cluster uses a single [GTID](../glossary.md#gtid) sequence. To make sure `GTID_PURGED` is the same, run the following commands on DC1-1, DC1-2, and DC1-3 nodes:

    ```bash
    mysql> flush logs;

>     mysql> purge source logs to 'mysql-bin.000003';

Reviewer 4: purge master logs to 'mysql-bin.000003';

Author: I have replaced `purge source logs` with `purge binary logs`

    ```

>     `mysql-bin.000003` is the last file returned from the show source logs. 

Reviewer 4: show master logs; 
Please mark it as a command
(https://dev.mysql.com/doc/refman/8.0/en/show-binary-logs.html)

Author: I have replaced show master logs with 

mysql>`SHOW BINARY LOGS`

as stated in https://dev.mysql.com/doc/refman/8.0/en/show-binary-logs.html

5. Adjust the aformentioned steps to set up the DC2 and DC3 clusters accordingly.
    
6. Add privileges for replication between all nodes only for DC1: 

    ```bash

>     GRANT REPLICATION CLIENT, REPLICATION REPLICA ON *.* TO 'repl'@'%' identified by 'replpass';

Reviewer 4: REPLICATION SLAVE

Author: Updated

    ```

7. Add the replication tables. For example, we add the tables to DC1: 
    
    ```text
    create database if not exists percona;
    use percona;
    CREATE TABLE `replication` (
      `host` varchar(40) NOT NULL,
      `weight` int(11) NOT NULL DEFAULT 0,
      `localIndex` int(11) DEFAULT NULL,

>       `isReplica` enum('No','Yes','Proposed','Failed') DEFAULT 'No',

Reviewer 4: The same 'isSlave'

Author: Updated

      `lastUpdate` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
      `lastHeartbeat` timestamp NOT NULL DEFAULT '1971-01-01 00:00:00',
      `connectionName` varchar(64) NOT NULL,
      PRIMARY KEY (`connectionName`,`host`),
      KEY `idx_host` (`host`)
    ) ENGINE=InnoDB DEFAULT CHARSET=latin1;
    CREATE TABLE `cluster` (
      `cluster` varchar(31) NOT NULL,
      `sourceCandidates` varchar(255) NOT NULL,
      `replCreds` varchar(255) NOT NULL,
      PRIMARY KEY (`cluster`)
    ) ENGINE=InnoDB DEFAULT CHARSET=latin1;
    CREATE TABLE `link` (
      `clusterReplica` varchar(31) NOT NULL,

>       `clusterSource` varchar(31) NOT NULL,

Reviewer 4: Please bring back 'clusterMaster' and 'clusterSlave'  instead of clusterSource and clusterReplica as this breaks the script

Author: Updated
